### PR TITLE
feat: add points

### PR DESCRIPTION
**Pull request by JoenvandeVorle**
## Description

Adds the point system to the game. You can now get a high score!!
Also adds point components to prefabs that need them.

## Testability

Open the `AddPoints` scene and interact with the objects there. Picking up potions should give points, as should killing an enemy.
The win screen should now display the points

## Checklist
- [x] Set the proper pull request name
- [x] Merged main into your branch
- [x] Added a scene to the game for your changes

### CODE REVIEW

**Review by JensSteenmetz — CHANGES_REQUESTED** [review]
Could you do the PointsAdderComponents similar to the EnemyPointsAdder? So for example, for the door, create an 'DoorOpened' event, that the DoorPointsAdder subscribes to.

File: aplib.net-demo/Assets/Scripts/Points/DoorPointsAdder.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+using Assets.Scripts.Doors;
+
+public class DoorPointsAdder : PointsAdderComponent
+{
+    private void Awake()
+    {
+        Door door = GetComponent<Door>();
+        door.DoorOpened += SendPoints;
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Points/DoorPointsAdder.cs:8]: You should be subscribing to and unsubscribing from events in OnEnable() and OnDisable.

File: aplib.net-demo/Assets/Scripts/Points/ItemPointsAdder.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+using Assets.Scripts.Items;
+using UnityEngine;
+
+public class ItemPointsAdder : PointsAdderComponent
+{
+    private void Awake()
+    {
+        PickupableItem item = GetComponent<PickupableItem>();
+        item.ItemPickedUp += SendPoints;
+    }
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Points/ItemPointsAdder.cs:10]: You should be subscribing to and unsubscribing from events in OnEnable() and OnDisable.

File: aplib.net-demo/Assets/Scripts/Points/KeyPointsAdder.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
+using Assets.Scripts.Items;
+using UnityEngine;
+
+public class KeyPointsAdder : PointsAdderComponent
+{
+    private void Awake()
+    {
+        PickupableKey item = GetComponent<PickupableKey>();
+        item.KeyPickedUp += SendPoints;
+    }
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Points/KeyPointsAdder.cs:10]: You should be subscribing to and unsubscribing from events in OnEnable() and OnDisable.

File: aplib.net-demo/Assets/Scripts/Points/KeyPointsAdder.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
+using Assets.Scripts.Items;
+using UnityEngine;
+
+public class KeyPointsAdder : PointsAdderComponent
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Points/KeyPointsAdder.cs:4]: RequireComponent

File: aplib.net-demo/Assets/Scripts/Points/ItemPointsAdder.cs [hunk 5]
```diff
@@ -0,0 +1,17 @@
+using Assets.Scripts.Items;
+using UnityEngine;
+
+public class ItemPointsAdder : PointsAdderComponent
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Points/ItemPointsAdder.cs:4]: RequireComponent

File: aplib.net-demo/Assets/Scripts/Points/DoorPointsAdder.cs [hunk 6]
```diff
@@ -0,0 +1,10 @@
+using Assets.Scripts.Doors;
+
+public class DoorPointsAdder : PointsAdderComponent
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Points/DoorPointsAdder.cs:3]: RequireComponent

File: aplib.net-demo/Assets/Scripts/Points/EnemyPointsAdder.cs [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    private HealthComponent _enemyHealth;
+
+    /// <summary>
+    /// Grab the enemy health component.
+    /// </summary>
+    private void Awake()
+    {
+        _enemyHealth = GetComponent<HealthComponent>();
+        _enemyHealth.Death += SendPoints;
+    }
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Points/EnemyPointsAdder.cs:15]: You should be subscribing to and unsubscribing from events in OnEnable() and OnDisable.

File: aplib.net-demo/Assets/Scripts/Items/PickupableItem.cs [hunk 8]
```diff
@@ -9,24 +9,36 @@ namespace Assets.Scripts.Items
...
     public class PickupableItem : MonoBehaviour
     {
-        private void OnTriggerEnter(Collider other)
-        {
-            if (!other.CompareTag("Player") || !other.material.name.Contains("PlayerPhysic"))
-                return;
+        /// <summary>
+        /// Event that is triggered when the item is picked up.
+        /// </summary>
+        public event System.Action<Item> ItemPickedUp;
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Items/PickupableItem.cs:15]: Nit: using System;

File: aplib.net-demo/Assets/Scripts/Items/PickupableKey.cs [hunk 9]
```diff
@@ -6,26 +6,34 @@ namespace Assets.Scripts.Items
     /// Attach this class to make object pickupable.
     /// </summary>
     [RequireComponent(typeof(Key))]
+    [RequireComponent(typeof(KeyPointsAdder))]
     public class PickupableKey : MonoBehaviour
     {
+        public event System.Action<Key> KeyPickedUp;
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Items/PickupableKey.cs:12]: Nit: using System;

File: aplib.net-demo/Assets/Scripts/Points/PointsManager.cs [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+    private static PointsManager _instance;
+
+    [SerializeField]
+    private readonly int _unusedItemPoints = 5;
+
+    /// <summary>
+    /// Private constructor to prevent instantiation by other classes.
+    /// </summary>
+    private PointsManager() { }
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Points/PointsManager.cs:42]: Do we need to enforce this? I'm not saying we don't, but I'm wondering about the reasoning.

**Review by JensSteenmetz — CHANGES_REQUESTED** [review]
A few more comments

File: aplib.net-demo/Assets/Scripts/Points/PointsManager.cs (hunk 10 above)
Comment (JoenvandeVorle) [reply aplib.net-demo/Assets/Scripts/Points/PointsManager.cs:42]: Yes, private constructors are part of the standard Singleton pattern iirc

**Review by Tboefijn — CHANGES_REQUESTED** [review]
Looks good! Left some comments.

File: aplib.net-demo/Assets/Scripts/Doors/Door.cs [hunk 11]
```diff
@@ -1,18 +1,18 @@
+using System;
 using TMPro;
 using UnityEngine;
+using Random = UnityEngine.Random;
```
[the remark this answers is not in the document]
Comment (JoenvandeVorle) [reply aplib.net-demo/Assets/Scripts/Doors/Door.cs:4]: gaat naar backlog!

**Review by JensSteenmetz — APPROVED** [review]
CID (consider it done)

**Review by Tboefijn — DISMISSED** [review]
LGTM!

**Review by Tboefijn — DISMISSED** [review]
LGTM!

**Review by Tboefijn — APPROVED** [review]
LGTM!

**Review by JensSteenmetz — APPROVED** [review]
LGTM

**Pull request merged**